core/src/main/java/com/linecorp/armeria/client/retry/RetryingRpcClient.java

```diff
@@ -0,0 +1,118 @@
+     * Creates a new {@link Client} decorator that handles failures of an invocation and retries RPC requests.
+     */
+    public static Function<Client<? super RpcRequest, ? extends RpcResponse>, RetryingRpcClient>
+    newDecorator(RetryRequestStrategy<RpcRequest, RpcResponse> retryRequestStrategy) {
+        return delegate -> new RetryingRpcClient(delegate, retryRequestStrategy,
+                                                 Backoff::withoutDelay);
+    }
+
+    /**
+     * Creates a new {@link Client} decorator that handles failures of an invocation and retries RPC requests.
+     */
+    public static Function<Client<? super RpcRequest, ? extends RpcResponse>, RetryingRpcClient>
+    newDecorator(RetryRequestStrategy<RpcRequest, RpcResponse> retryRequestStrategy,
+                 Supplier<? extends Backoff> backoffSupplier) {
+        return delegate -> new RetryingRpcClient(delegate, retryRequestStrategy, backoffSupplier);
+    }
+
+    /**
+     * Creates a new instance that decorates the specified {@link Client}.
+     */
+    public RetryingRpcClient(Client<? super RpcRequest, ? extends RpcResponse> delegate,
+                             RetryRequestStrategy<RpcRequest, RpcResponse> retryStrategy,
+                             Supplier<? extends Backoff> backoffSupplier) {
+        super(delegate, retryStrategy, backoffSupplier);
+    }
+
+    @Override
+    public RpcResponse execute(ClientRequestContext ctx, RpcRequest req) throws Exception {
+        DefaultRpcResponse responseFuture = new DefaultRpcResponse();
+        Backoff backoff = newBackoff();
+        retry(0, backoff, ctx, req, () -> {
+            try {
+                return delegate().execute(ctx, req);
+            } catch (Exception e) {
+                return new DefaultRpcResponse(e);
+            }
+        }, responseFuture);
+        return responseFuture;
+    }
+
+    private void retry(int numAttemptsSoFar, Backoff backoff, ClientRequestContext ctx, RpcRequest req,
+                       Supplier<RpcResponse> action, DefaultRpcResponse responseFuture) {
+        RpcResponse response = action.get();
+        response.handle(voidFunction((result, thrown) -> {
+            final Throwable exception;
+            if (thrown != null) {
+                if (!retryStrategy().shouldRetry(req, thrown)) {
+                    responseFuture.completeExceptionally(thrown);
+                    return;
+                }
+                exception = thrown;
+            } else if (retryStrategy().shouldRetry(req, response)) {
+                exception = RETRY_EXCEPTION;
+            } else {
+                responseFuture.complete(result);
+                return;
+            }
+
+            long nextInterval = backoff.nextIntervalMillis(numAttemptsSoFar);
+            if (nextInterval < 0) {
+                responseFuture.completeExceptionally(exception);
+            } else {
+                EventLoop eventLoop = ctx.contextAwareEventLoop().next();
```

Comment:
Don't want to call ```.next()``` I think, we generally don't want to move requests among event loops. In practice, I guess the context's event loop is a single-threaded loop where next doesn't do anything anyways.

Also, the next() actually loses the context (I guess it may be a bug in RequestContextAwareEventLoop where next() should wrap the result in a ContextAwareEventLoop);